Library/Homebrew/cmd/pull.rb

```diff
@@ -202,81 +217,68 @@ def pull
-        safe_system "git", "merge", "--ff-only", "--no-edit", bottle_branch
-        safe_system "git", "branch", "-D", bottle_branch
+        safe_system "git", "checkout", "--quiet", "-B", bottle_branch, orig_revision
+        pull_patch bottle_commit_url, "bottle commit"
+        safe_system "git", "rebase", "--quiet", branch
+        safe_system "git", "checkout", "--quiet", branch
+        safe_system "git", "merge", "--quiet", "--ff-only", "--no-edit", bottle_branch
+        safe_system "git", "branch", "--quiet", "-D", bottle_branch
 
         # Publish bottles on Bintray
-        bintray_user = ENV["BINTRAY_USER"]
-        bintray_key = ENV["BINTRAY_KEY"]
-
-        if bintray_user && bintray_key
-          repo = Bintray.repository(tap)
-          changed_formulae.each do |f|
-            next if f.bottle_unneeded? || f.bottle_disabled?
-            ohai "Publishing on Bintray:"
-            package = Bintray.package f.name
-            version = f.pkg_version
-            curl "-w", '\n', "--silent", "--fail",
-              "-u#{bintray_user}:#{bintray_key}", "-X", "POST",
-              "-H", "Content-Type: application/json",
-              "-d", '{"publish_wait_for_secs": 0}',
-              "https://api.bintray.com/content/homebrew/#{repo}/#{package}/#{version}/publish"
-            bintray_fetch_formulae << f
-          end
-        else
-          opoo "You must set BINTRAY_USER and BINTRAY_KEY to add or update bottles on Bintray!"
+        unless ARGV.include? "--no-publish"
+          published = publish_changed_formula_bottles(tap, changed_formulae)
+          bintray_published_formulae.concat(published).uniq!
         end
       end
 
       ohai "Patch changed:"
-      safe_system "git", "diff-tree", "-r", "--stat", revision, "HEAD"
+      safe_system "git", "diff-tree", "-r", "--stat", orig_revision, "HEAD"
 
-      if ARGV.include? "--install"
-        changed_formulae.each do |f|
-          ohai "Installing #{f.full_name}"
-          install = f.installed? ? "upgrade" : "install"
-          safe_system "brew", install, "--debug", f.full_name
-        end
-      end
+      all_changed_formulae.concat(changed_formulae).uniq!
     end
 
-    bintray_fetch_formulae.each do |f|
-      max_retries = 8
-      retry_count = 0
-      begin
-        success = system "brew", "fetch", "--force-bottle", f.full_name
-        raise "Failed to download #{f} bottle!" unless success
-      rescue RuntimeError => e
-        retry_count += 1
-        raise e if retry_count >= max_retries
-        sleep_seconds = 2**retry_count
-        ohai "That didn't work; sleeping #{sleep_seconds} seconds and trying again..."
-        sleep sleep_seconds
-        retry
-      end
-    end
+    # Verify bintray publishing after all patches have been applied
+    verify_bintray_published(bintray_published_formulae)
+  end
+
+  def force_utf8!(str)
+    str.force_encoding("UTF-8") if str.respond_to?(:force_encoding)
   end
```

Comment:
Similar to my proposal on other PR for `ruby2orabove?` method, I think we should create similar global method for encoding fix.